### PR TITLE
feat: Create platform abstraction traits (Issue #94)

### LINKED ISSUE #94: Create platform abstraction traits
## Summary
Create trait definitions for platform abstraction layer to enable cross-platform support.

**Priority**: 🔴 Critical
**Phase**: 1 - Foundation

## Tasks
- [ ] Create `src/platform/traits.rs` with `InputBlocker`, `PowerManager`, `PermissionChecker`, `SystemTray`, `OverlayWindow` traits
- [ ] Create `src/platform/types.rs` with `KeyEvent`, `Modifiers`, `SleepAssertion`, `Rect` types
- [ ] Define error types: `InputBlockError`, `PowerError`, `TrayError`, `WindowError`

## Acceptance Criteria
- [ ] All traits compile
- [ ] Types are platform-agnostic (no OS-specific imports)

## Technical Details
```rust
pub trait InputBlocker: Send + Sync {
    fn setup(&mut self) -> Result<(), InputBlockError>;
    fn disable(&mut self);
}

pub trait PowerManager: Send + Sync {
    fn prevent_sleep(&self) -> Result<SleepAssertion, PowerError>;
    fn allow_sleep(&self, assertion: SleepAssertion) -> Result<(), PowerError>;
}

pub trait PermissionChecker {
    fn check_permissions(&self) -> bool;
    fn check_permissions_with_prompt(&self) -> bool;
    fn open_permissions_settings(&self) -> bool;
}

pub trait SystemTray {
    fn setup(&mut self) -> Result<(), TrayError>;
    fn set_state(&mut self, active: bool);
}

pub trait OverlayWindow {
    fn create(&mut self) -> Result<(), WindowError>;
    fn show(&mut self);
    fn hide(&mut self);
    fn set_opacity(&mut self, opacity: f64);
}
```

### PR DESCRIPTION
**Pull request by taearls**
## Summary

Create trait definitions for platform abstraction layer to enable cross-platform support for Windows and Linux.

This is the critical foundation issue for Phase 8: Cross-Platform Support. All other platform-specific issues (#95-#113) depend on this.

### New Files

**`src/platform/types.rs`** - Platform-agnostic types:
- `Modifiers`: Keyboard modifier keys (command, option, control, shift)
- `KeyEvent`: Keyboard events with keycode, modifiers, and key state
- `SleepAssertion`: Opaque handle for sleep prevention assertions
- `Rect`: Rectangle type for window positioning and sizing

**`src/platform/errors.rs`** - Error types for each platform component:
- `InputBlockError`: For input blocking operations
- `PowerError`: For power management operations
- `TrayError`: For system tray operations
- `WindowError`: For overlay window operations
- `PermissionError`: For permission checking operations

**`src/platform/traits.rs`** - Platform abstraction traits:
- `InputBlocker`: Intercept and block keyboard input (macOS: CGEventTap, Windows: keyboard hooks, Linux: X11 grab)
- `PowerManager`: Prevent system sleep during protection (macOS: IOKit, Windows: SetThreadExecutionState, Linux: D-Bus)
- `PermissionChecker`: Handle platform permission requirements (macOS: Accessibility)
- `SystemTray`: Manage menu bar / system tray icon
- `OverlayWindow`: Create and manage fullscreen overlay

### Updated Files

- **`src/platform/mod.rs`**: Export new modules and re-export commonly used types
- **`ROADMAP.md`**: Mark issue #94 as complete, update counts and critical path

## Technical Notes

- All traits include comprehensive documentation with platform-specific implementation notes
- `InputBlocker` and `PowerManager` traits require `Send + Sync` for thread safety
- Error types implement `std::error::Error` and `std::fmt::Display`
- Types use `#[must_use]` and `const fn` where appropriate

## Test Plan

- [x] All 259 tests pass (27 new tests added)
- [x] `cargo clippy` clean
- [x] `cargo fmt --check` passes
- [x] `cargo build --release` succeeds

## Unblocks

This PR unblocks the following issues:
- #95: Reorganize macOS platform code into subdirectory
- #97: Create canonical Key enum and split keycodes by platform
- #99: Update Cargo.toml for conditional platform dependencies
- #100, #101: Windows implementations (partially)
- #106, #107, #108: Linux implementations (partially)

Closes #94

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Platform abstraction completed: cross-platform interfaces for input blocking, power management, permissions, system tray, and overlay windows.
* **Chores**
  * Added platform-agnostic types (key events, modifiers, rects, sleep assertions) and restructured platform surface for future ports.
* **Tests**
  * Comprehensive unit tests validating new types, traits, and error formatting.
* **Documentation**
  * Roadmap and changelog updated to reflect completed foundation work and reordered milestones.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Review by taearls — COMMENTED** [review]
## Summary

This PR creates the foundation for cross-platform support by defining platform abstraction traits, types, and error types. This is the critical foundation issue (#94) that unblocks all other platform-specific work (#95-#113).

**Changed files**: 5 files, +1059 additions, -22 deletions
**Impact areas**: Platform abstraction layer, cross-platform architecture
**Review depth**: Full validation

## Quality Checks Results

- ✅ Format checking: `cargo fmt --check` - Pass
- ✅ Linting: `cargo clippy -- -D warnings` - Pass
- ✅ Tests: `cargo test` - 259 passed, 0 failed
- ✅ Build: `cargo build --release` - Pass

## Code Review Findings

#### ✅ Positive Observations

1. **Excellent Documentation**
   - All traits have comprehensive rustdoc with platform-specific implementation notes
   - Module-level documentation explains the architecture clearly
   - Each method has documented behavior, errors, and examples where appropriate
   - `src/platform/traits.rs:1-19` - Great module overview

2. **Idiomatic Rust**
   - Proper use of `#[must_use]` attributes on all constructors and getters
   - `const fn` used where possible for compile-time evaluation
   - Correct derive macros (`Debug`, `Clone`, `Copy`, `PartialEq`, `Eq`)
   - `src/platform/types.rs:33-36` - Good use of const fn

3. **Thread Safety Design**
   - `InputBlocker` and `PowerManager` correctly require `Send + Sync`
   - Clear documentation explaining why thread safety is required
   - `src/platform/traits.rs:67-70` and `src/platform/traits.rs:117-121`

4. **Comprehensive Error Types**
   - Each platform component has its own error enum
   - Error variants cover all expected failure modes
   - All errors implement `std::error::Error` and `Display`
   - Platform-specific escape hatch (`Platform(String)`) in each error type

5. **Thorough Testing**
   - 27 new unit tests covering types, errors, and trait mocks
   - Tests verify Send + Sync bounds with compile-time checks
   - Good edge case coverage (e.g., opacity clamping, empty rects)
   - Mock implementations demonstrate how traits should be used

6. **Clean Module Organization**
   - Re-exports commonly used types at module level
   - Clear separation: types, errors, traits
   - Forward-looking comment about future reorganization (issue #95)

#### 🔵 Minor Issues / Suggestions

1. **Consider `thiserror` for Error Implementations** (Optional)
   - **File**: `src/platform/errors.rs`
   - **Issue**: Manual `impl Display` and `impl Error` for each error type is boilerplate
   - **Suggestion**: Consider using `thiserror` crate for more concise error definitions
   - **Note**: Current implementation is fine - this is just a potential future improvement

2. **Rect could use Copy for operations** (Optional)
   - **File**: `src/platform/types.rs:130-144`
   - **Issue**: Methods like `is_empty()` and `area()` take `&self` but `Rect` is `Copy`
   - **Note**: Current API is idiomatic and efficient - not a problem

### Testing Analysis

- **Coverage**: Excellent - all new types, errors, and trait behavior tested
- **Test levels**: Unit tests appropriate for trait definitions and types
- **Edge cases**: Well covered (empty rects, opacity bounds, modifier combinations)
- **Test quality**: Clear assertions, no flaky tests, well-organized

### Architecture & Patterns Compliance

- **Follows documented architecture**: Yes - extends platform module appropriately
- **Consistent with codebase patterns**: Yes - matches existing module structure
- **Separation of concerns**: Excellent - types, errors, and traits cleanly separated
- **Design patterns**: Trait-based abstraction is appropriate for cross-platform code

### Security Review

- ✅ No exposed secrets in code or commit history
- ✅ No unsafe code in new files (traits are implemented safely)
- ✅ Error messages don't leak sensitive information
- N/A - No network, file, or database operations in this PR

### Performance Review

- **Algorithmic efficiency**: All operations are O(1)
- **Resource management**: Types are lightweight (Copy/Clone)
- **Memory**: No allocations in hot paths (except error messages)
- **Zero-cost abstractions**: Traits will be monomorphized at compile time

### Documentation Review

- ✅ Comprehensive rustdoc on all public items
- ✅ Module-level documentation explains architecture
- ✅ Platform-specific implementation notes in trait docs
- ✅ Examples in doc comments (marked `ignore` for code that can't run in tests)
- ✅ ROADMAP.md updated with completion status and changelog

## Approval Status

💬 **COMMENT** - This is excellent foundational work. The code is well-documented, thoroughly tested, and follows Rust best practices. Ready to merge. (Self-review: cannot approve own PR)

**Reasoning**: 
- All quality checks pass (format, clippy, tests, build)
- Excellent documentation and test coverage
- Clean, idiomatic Rust code
- No critical or major issues found
- Properly unblocks dependent issues (#95-#113)

---

**Review completed using**: cargo fmt --check, cargo clippy -- -D warnings, cargo test, cargo build --release
**Reviewed by**: Claude Code Agent

File: src/platform/types.rs [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+        }
+        if self.control {
+            parts.push("Ctrl");
+        }
+        if self.shift {
+            parts.push("Shift");
+        }
+        write!(f, "{}", parts.join("+"))
+    }
+}
```
[the remark this answers is not in the document]
Comment (taearls) [reply src/platform/types.rs]: Fixed in commit b873315. The `Display` impl now returns "None" when no modifiers are pressed instead of an empty string. Added a test to verify this behavior.

File: src/platform/types.rs [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+    #[must_use]
+    pub fn is_empty(&self) -> bool {
+        self.width <= 0.0 || self.height <= 0.0
+    }
+
+    /// Returns the area of the rectangle.
+    #[must_use]
+    pub fn area(&self) -> f64 {
+        self.width * self.height
+    }
```
Comment (taearls) [inline src/platform/types.rs]: Fixed in commit b873315. `Rect::area()` now returns 0.0 for empty or invalid rectangles (those with zero or negative dimensions). Added comprehensive tests for this edge case.
